sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -611,6 +611,29 @@ object ColumnPruning extends Rule[LogicalPlan] {
     // Can't prune the columns on LeafNode
     case p @ Project(_, _: LeafNode) => p
 
+    // If the current project or the child references to nested fields, we can substitute them
+    // by alias attributes; then a project of the nested fields as aliases on the children
+    // of the child will be created.
+    //
+    // Note that if the child is also a [[Project]], it will be handled by the previous rule.
+    // If the child is a [[Filter]], it will be conflict with PushPredicatesThroughProject.
+    // When the child is a [[SerializeFromObject]], we can not push the projection through it.
+    case p @ Project(_, child) if !child.isInstanceOf[Project] && !child.isInstanceOf[Filter] &&
```

Comment:
I am generally not a big fan of blacklisting, we have seen various optimizer bugs caused by blacklisting in the past. Can we just create a white list of operators?

Reply:
I was debating the same. Will change to a white list of operators.